sw/device/silicon_creator/lib/sigverify/sphincsplus/hash_shake.c

```diff
@@ -0,0 +1,88 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+//
+// Derived from code in the SPHINCS+ reference implementation (CC0 license):
+// https://github.com/sphincs/sphincsplus/blob/ed15dd78658f63288c7492c00260d86154b84637/ref/hash_shake.c
+
+#include "sw/device/lib/base/memory.h"
+#include "sw/device/silicon_creator/lib/drivers/kmac.h"
+#include "sw/device/silicon_creator/lib/sigverify/sphincsplus/address.h"
+#include "sw/device/silicon_creator/lib/sigverify/sphincsplus/hash.h"
+#include "sw/device/silicon_creator/lib/sigverify/sphincsplus/params.h"
+#include "sw/device/silicon_creator/lib/sigverify/sphincsplus/utils.h"
+
+enum {
+  /**
+   * Number of bits needed to represent the `tree` field.
+   */
+  kSpxTreeBits = kSpxTreeHeight * (kSpxD - 1),
+  /**
+   * Number of bytes needed to represent the `tree` field.
+   */
+  kSpxTreeBytes = (kSpxTreeBits + 7) / 8,
+  /**
+   * Number of bits needed to represent a leaf index.
+   */
+  kSpxLeafBits = kSpxTreeHeight,
+  /**
+   * Number of bytes needed to represent a leaf index.
+   */
+  kSpxLeafBytes = (kSpxLeafBits + 7) / 8,
+  /**
+   * Number of bytes needed for the message digest.
+   */
+  kSpxDigestBytes = kSpxForsMsgBytes + kSpxTreeBytes + kSpxLeafBytes,
+  /**
+   * Number of 32-bit words needed for the message digest.
+   *
+   * Rounded up if necessary.
+   */
+  kSpxDigestWords = (kSpxDigestBytes + sizeof(uint32_t) - 1) / sizeof(uint32_t),
+};
+
+static_assert(
+    kSpxTreeBits <= 64,
+    "For given height and depth, 64 bits cannot represent all subtrees.");
+
+rom_error_t spx_hash_initialize(spx_ctx_t *ctx) {
+  return kmac_shake256_configure();
+}
+
+rom_error_t spx_hash_message(const unsigned char *R, const unsigned char *pk,
+                             const unsigned char *m, unsigned long long mlen,
+                             const spx_ctx_t *ctx, unsigned char *digest,
+                             uint64_t *tree, uint32_t *leaf_idx) {
+  // Suppress a warning about this variable being unused. For SHAKE it's
+  // unnecessary, but it is included in the interface because other SPHINCS+
+  // hash functions may need it.
+  (void)ctx;
```

Comment:
I think it is also useful to point out why we have this in the function definition as well and add this to the documentation.  If it is not really needed, then my recommendation would be to remove it.